staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -220,31 +224,50 @@ func NewSharedInformerWithOptions(lw ListerWatcher, exampleObject runtime.Object
 // requested before the informer starts and the
 // defaultEventHandlerResyncPeriod given here and (b) the constant
 // `minimumResyncPeriod` defined in this file.
-func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defaultEventHandlerResyncPeriod time.Duration, indexers Indexers) SharedIndexInformer {
-	opts := InformerOptions{
-		ResyncPeriod: defaultEventHandlerResyncPeriod,
-		Indexers:     indexers,
-	}
-	return NewSharedIndexInformerWithOptions(lw, exampleObject, opts)
-}
-
-//TODO(KCP)
-func NewSharedIndexInformerWithOptions(lw ListerWatcher, exampleObject runtime.Object, opts InformerOptions) SharedIndexInformer {
+func NewSharedIndexInformerWithOptions(lw ListerWatcher, exampleObject runtime.Object, opts ...SharedInformerOption) SharedIndexInformer {
 	realClock := &clock.RealClock{}
 	sharedIndexInformer := &sharedIndexInformer{
-		processor:                       &sharedProcessor{clock: realClock},
-		keyFunc:                         opts.KeyFunction,
-		indexer:                         NewIndexer(opts.KeyFunction, opts.Indexers),
-		listerWatcher:                   lw,
-		objectType:                      exampleObject,
-		resyncCheckPeriod:               opts.ResyncPeriod,
-		defaultEventHandlerResyncPeriod: opts.ResyncPeriod,
-		cacheMutationDetector:           NewCacheMutationDetector(fmt.Sprintf("%T", exampleObject)),
-		clock:                           realClock,
+		processor:             &sharedProcessor{clock: realClock},
+		listerWatcher:         lw,
+		objectType:            exampleObject,
+		cacheMutationDetector: NewCacheMutationDetector(fmt.Sprintf("%T", exampleObject)),
+		clock:                 realClock,
+	}
+
+	for _, opt := range opts {
+		opt(sharedIndexInformer)
+	}
+
+	if sharedIndexInformer.keyFunc == nil {
+		WithKeyFunction(MetaNamespaceKeyFunc)(sharedIndexInformer)
 	}
+
 	return sharedIndexInformer
 }
 
+type SharedInformerOption func(*sharedIndexInformer)
+
+func WithResyncPeriod(t time.Duration) SharedInformerOption {
+	return func(s *sharedIndexInformer) {
+		s.resyncCheckPeriod = t
+		s.defaultEventHandlerResyncPeriod = t
+	}
+}
+
+func WithKeyFunction(k KeyFunc) SharedInformerOption {
+	return func(s *sharedIndexInformer) {
+		s.keyFunc = k
+		s.indexer = NewIndexer(s.keyFunc, s.indexers)
+	}
+}
+
+func WithIndexers(ind Indexers) SharedInformerOption {
+	return func(s *sharedIndexInformer) {
+		s.indexers = ind
+		s.indexer = NewIndexer(s.keyFunc, s.indexers)
```

Comment:
What happens if `WithIndexers` is invoked **before** `WithKeyFunction`? I think there may be an ordering issue here. Could we add a `complete` function to finalize dependent fields **after** all of the options have been applied?

e.g.

```go
// In NewSharedIndexInformerWithOptions(...)
for _, opt := range opts {
  opt(sharedIndexInformer)
}

complete(sharedIndexInformer)

// ...
```

Reply:
I didn't make it a function but I did basically do this in the followup commit, let me go ahead and squash these to ease review

Reply:
https://github.com/kcp-dev/kubernetes/pull/59/commits/c1d54e28a27cb66ffa0181620b7d8c70f319e8f4